feat: Add blogpost for 1.12 release (#221)
kubewarden/kubewarden.io: content/blog/2024/04/kubewarden-1-12-release.md

---
title: Kubewarden 1.12 release
authors:
  - Víctor Cuadrado
date: 2024-04-24
---

Today we're glad to announce the release of Kubewarden 1.12.

This release focuses on optimizations and high availability, both oriented to
production.

## Optimizing Gatekeeper policies

The previous [1.11
release](https://www.kubewarden.io/blog/2024/03/kubewarden-1-11-release/)
featured lots of optimizations for
[context aware policies](https://docs.kubewarden.io/next/reference/spec/context-aware-policies).

The 1.12 release provides a further optimization for Gatekeeper policies that
access Kubernetes resources. This optimization provides an extra 55%
performance boost for these policies.

> This is particularly noticeable on really big clusters (thousands of nodes), as
> Gatekeeper policies expect a serialized inventory of all Kubernetes objects
> they could access, under `data.inventory`.

Reviewer: ```suggestion
The benefits of this optimization are particularly noticeable when a huge number
of Kubernetes resources are accessed by a Gatekeeper policy.

For example, a policy that access Pod resources, will see a reduced evaluation time
on big clusters where hundreds of Pod objects are defined.
```

> In addition, Gatekeeper policies in Kubewarden have `spec.contextAwareResources`
> which allows us to fine-tune the context-aware permissions per policy.
> Leveraging this feature allows us to provide greater cache granularity for each
> resource set shared between policies (e.g: "all Namespaces with a specific
> LabelSelector").

Reviewer: ```suggestion
```

I would drop this section. Gatekeeper doesn't use label selectors when accessing resources, hence we cannot filter that.

What we have always been doing is to expose to the Gatekeeper policy only the resources it's interested about.

With Gatekeeper the admin determines what has to be shared with policies, then this is provided to **all** policies, even if they do not care.

---
title: Kubewarden 1.12 release
authors:
  - Víctor Cuadrado
date: 2024-04-24
---

Today we're glad to announce the release of Kubewarden 1.12.

This release focuses on optimizations and High-Availability, both oriented to
production.

## Optimizing Rego policies

On really big clusters (thousands of nodes), Gatekeeper policies that make use
of context-aware data are markedly slower than those written in Go/Rust. This
is because Gatekeeper policies expect a serialized inventory of all Kubernetes
objects they could access under `data.inventory`, for which the policy-server

> must spent significant CPU cycles.

Author: This block changed; I have applied some of its suggestions by hand.

Gatekeeper policies in Kubewarden in addition have
`spec.contextAwareResources`, a Kubewarden feature that allows us to fine-tune

> the context-aware permissions per policy, and provide more security.

Author: This block changed; I have applied some of its suggestions by hand.

Following requests from the community, for Kubewarden 1.12 we have elected to
cache this `data.inventory` of Gatekeeper policies for performance. The cache,
thanks to `spec.contextAwareResources`, has an increased granularity compared

> to vanilla Gatekeeper.

Author: This block changed; I have applied some of its suggestions by hand.

The cache works with sets of resources as defined in the policy rules and
`spec.contextAwareResources`. Hence policies that access the same set of
resources (e.g: all Namespaces with a specific LabelSelector) will share the
same inventory. That cache set is invalidated whenever a resource of that set

> changes.

Author: This block changed; I have applied some of its suggestions by hand.

> On really big clusters (thousands of nodes), Gatekeeper policies that make use
> of context-aware data are markedly slower than those written in Go/Rust. This
> is because Gatekeeper policies expect a serialized inventory of all Kubernetes
> objects they could access under `data.inventory`, for which the policy-server
> must spent significant CPU cycles.
> 
> Gatekeeper policies in Kubewarden in addition have
> `spec.contextAwareResources`, a Kubewarden feature that allows us to fine-tune
> the context-aware permissions per policy, and provide more security.
> 
> Following requests from the community, for Kubewarden 1.12 we have elected to
> cache this `data.inventory` of Gatekeeper policies for performance. The cache,
> thanks to `spec.contextAwareResources`, has an increased granularity compared
> to vanilla Gatekeeper.
> 
> The cache works with sets of resources as defined in the policy rules and
> `spec.contextAwareResources`. Hence policies that access the same set of
> resources (e.g: all Namespaces with a specific LabelSelector) will share the
> same inventory. That cache set is invalidated whenever a resource of that set
> changes.
> 
> Stay tuned for more performance improvements on next releases!

Reviewer: It's a bit too low level.

```suggestion
The previous [1.11 release](https://www.kubewarden.io/blog/2024/03/kubewarden-1-11-release/) featured lots of optimizations for
[context aware policies](https://docs.kubewarden.io/next/reference/spec/context-aware-policies).

The 1.12 release provides a further optimization for Gatekeeper policies that access Kubernetes resources.
This optimization provides an extra 55% performance boost for these policies.
```
